runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -270,8 +294,22 @@ public void setKeyContextElement1(StreamRecord record) throws Exception {
 
   @Override
   public void setCurrentKey(Object key) {
-    throw new UnsupportedOperationException(
-        "Current key for state backend can only be set by state requests from SDK workers.");
+    if (stateful && usesTimers) {
```

Comment:
I think it would be more readable to have an outer `if (usesTimers)` and then a nested `if (stateful)` - but how does synchronizing just `setCurrentKey` help, how is the current key then used (not synchronized?) when timers are present?

Reply:
Due to the concurrent access the synchronization is actually quite hard. I think I have pushed a good solution now.